### PR TITLE
Change the interfaces of builtinConcat to datums

### PR DESCRIPTION
**Pull request by zxylvlp**
Change the interfaces of builtinConcat to datums.

### CODE REVIEW

File: util/types/datum.go [hunk 1]
```diff
@@ -112,9 +112,13 @@ func (d *Datum) GetString() string {
 // SetString sets string value.
 func (d *Datum) SetString(s string) {
 	d.k = KindString
+	sink(s)
```
Comment (coocood) [inline util/types/datum.go]: Add some comment for this trick.

File: util/types/datum.go [hunk 2]
```diff
@@ -1166,6 +1170,40 @@ func (d *Datum) ToFloat64() (float64, error) {
 	}
 }
 
+// ToString converts an interface to a string.
```
Comment (shenli) [inline util/types/datum.go]: The comment is not clear.
"ToString gets the string representation of the datum" is better?

File: evaluator/builtin_string.go [hunk 3]
```diff
@@ -46,20 +46,23 @@ func builtinLength(args []interface{}, _ context.Context) (v interface{}, err er
...
 	for _, a := range args {
-		if a == nil {
-			return nil, nil
+		if a.Kind() == types.KindNull {
+			d.SetNull()
+			return d, nil
 		}
-		ss, err := types.ToString(a)
+		var ss string
+		ss, err = a.ToString()
```
Comment (zimulala) [inline evaluator/builtin_string.go]: It can remove line 56, and use `ss, err := a.ToString()` directly?

File: util/types/datum.go [hunk 4]
```diff
@@ -1166,6 +1170,40 @@ func (d *Datum) ToFloat64() (float64, error) {
...
+	case KindMysqlHex:
+		return d.GetMysqlHex().ToString(), nil
+	case KindMysqlBit:
+		return d.GetMysqlBit().ToString(), nil
+	case KindMysqlEnum:
+		return d.GetMysqlEnum().String(), nil
+	case KindMysqlSet:
+		return d.GetMysqlSet().String(), nil
+	default:
+		return "", errors.Errorf("cannot convert %v(type %T) to string", d.GetValue(), d.GetValue())
```
Comment (zimulala) [inline util/types/datum.go]: It can use the `invalidConv` function directly ?

Comment (zxylvlp) [inline util/types/datum.go]: The invalidConv return type is datum not string, it was used by datum to datum convertion, not datum to native type variable.

File: evaluator/builtin_string.go (hunk 3 above)
Comment (zxylvlp) [inline evaluator/builtin_string.go]: In that way, err would be shadowed.

**Comment by zxylvlp** [comment]
@zimulala @coocood @shenli

File: util/types/datum.go [hunk 5]
```diff
@@ -112,10 +112,12 @@ func (d *Datum) GetString() string {
 // SetString sets string value.
 func (d *Datum) SetString(s string) {
 	d.k = KindString
+	// Avoid s was allocated in stack.
 	sink(s)
 	d.b = hack.Slice(s)
 }
 
+// sink let s was allocated in heap instead of stack in the stage of escape analysis.
```
Comment (coocood) [inline util/types/datum.go]: prevents s from being allocated on the stack.

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**